internal/server/handlers.go

```diff
@@ -22,13 +24,61 @@ import (
 	"github.com/infrahq/infra/internal/server/models"
 	"github.com/infrahq/infra/internal/server/providers"
 	"github.com/infrahq/infra/internal/server/redis"
+	"github.com/infrahq/infra/uid"
 )
 
 type API struct {
 	t          *Telemetry
 	server     *Server
-	migrations []apiMigration
 	openAPIDoc openapi3.T
+	versions   map[routeIdentifier][]routeVersion
+}
+
+type routeVersion struct {
+	version *semver.Version
+	handler func(c *gin.Context)
+}
+
+func addVersionHandler[Req, Res any](a *API, method, path, version string, routeDef route[Req, Res]) {
+	if a.versions == nil {
+		a.versions = make(map[routeIdentifier][]routeVersion)
+	}
+
+	routeDef.routeSettings.omitFromDocs = true
+
+	key := routeIdentifier{method: method, path: path}
+	a.versions[key] = append(a.versions[key], routeVersion{
+		version: semver.MustParse(version),
+		handler: func(c *gin.Context) {
+			wrapped := wrapRoute(a, key, routeDef)
+			if err := wrapped(c); err != nil {
+				sendAPIError(c, err)
+			}
+		},
+	})
+}
+
+func (a *API) addPreviousVersionHandlers() {
+	type listAccessKeysRequest_v0_16_1 struct {
+		UserID      uid.ID `form:"user_id"`
+		Name        string `form:"name"`
+		ShowExpired bool   `form:"show_expired"`
+		api.PaginationRequest
+	}
+	addVersionHandler(a,
+		http.MethodGet, "/api/access-keys", "0.16.1",
+		route[listAccessKeysRequest_v0_16_1, *api.ListResponse[api.AccessKey]]{
+			routeSettings: defaultRouteSettingsGet,
+			handler: func(c *gin.Context, reqOld *listAccessKeysRequest_v0_16_1) (*api.ListResponse[api.AccessKey], error) {
+				req := &api.ListAccessKeysRequest{
```

Comment:
I _think_ you can just do `req := reqOld` here since the structs are the same.

Reply:
I think we'd need to type convert it as well. I decided against that approach because it will break as soon as we add another field to `api.ListAcessKeysRequest`. Someone adding a new field to that struct shouldn't need to come fix this code.